src/servers/src/http/jaeger.rs

```diff
@@ -0,0 +1,1244 @@
+                        value: Value::String(value.to_string()),
+                    }),
+                    JsonValue::Number(value) => Some(KeyValue {
+                        key,
+                        value_type: ValueType::Int64,
+                        value: Value::Int64(value.as_i64().unwrap_or(0)),
+                    }),
+                    JsonValue::Bool(value) => Some(KeyValue {
+                        key,
+                        value_type: ValueType::Boolean,
+                        value: Value::Boolean(value),
+                    }),
+                    // FIXME(zyy17): Do we need to support other types?
+                    _ => {
+                        warn!("Unsupported value type: {:?}", value);
+                        None
+                    }
+                })
+                .collect();
+            span.tags = tags;
+        }
+
+        if let Some(spans) = trace_id_to_spans.get_mut(&span.trace_id) {
+            spans.push(span);
+        } else {
+            trace_id_to_spans.insert(span.trace_id.clone(), vec![span]);
+        }
+    }
+
+    let mut traces = Vec::new();
+    for (trace_id, spans) in trace_id_to_spans {
+        let mut trace = Trace {
+            trace_id,
+            spans,
+            ..Default::default()
+        };
+
+        if let Some(processes) = trace_id_to_processes.remove(&trace.trace_id) {
+            let mut process_id_to_process = HashMap::new();
+            for (service_name, process_id) in processes.into_iter() {
+                process_id_to_process.insert(
+                    process_id,
+                    Process {
+                        service_name,
+                        tags: vec![],
+                    },
+                );
+            }
+            trace.processes = Some(process_id_to_process);
+        }
+        traces.push(trace);
+    }
+
+    traces
+}
+
+// Construct Jaeger operations from records.
+fn operations_from_records(records: HttpRecordsOutput, contain_span_kind: bool) -> Vec<Operation> {
+    let mut operations = Vec::with_capacity(records.total_rows);
+    for row in records.rows.into_iter() {
+        let mut row_iter = row.into_iter();
+        if let Some(JsonValue::String(operation)) = row_iter.next() {
+            let mut operation = Operation {
+                name: operation,
+                span_kind: None,
+            };
+            if contain_span_kind {
+                if let Some(JsonValue::String(span_kind)) = row_iter.next() {
+                    operation.span_kind = Some(normalize_span_kind(span_kind.as_str()).to_string());
+                }
+            } else {
+                // skip span kind.
+                row_iter.next();
+            }
+            operations.push(operation);
+        }
+    }
+    operations
+}
+
+// FIXME(zyy17): It may encounter performance issue when the trace table is too large.
+fn get_services_sql(db: &str, trace_table_name: &str) -> String {
+    format!(
```

Comment:
This will introduce SQL injection. I think we should use `DataFrame` instead.

Just like https://github.com/GreptimeTeam/greptimedb/blob/88c3d331a1039832e8cbe4e1ac71147a687e5c33/src/query/src/sql.rs#L222

Reply:
I use the DataFrame APIs to refactor all the query process. It may not have SQL injection problem in theory.